# Coursera Project On Toronto Neighborhood Segmentation

### Will showcase web-scraping and geo-spatial clustering

In [12]:
#importing libraries that will be used
import pandas as pd #data analysis
#ensure that we can print out an entire dataset if desired
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
import requests #handle web requests
from bs4 import BeautifulSoup #parse the resultant html

In [13]:
wikiurl = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M' #link containing our data to be analyzed
table_class = "wikitable sortable jquery-tablesorter" #define type of table expected to be collected from target link
response = requests.get(wikiurl) #retrieve html data from target link
print(response.status_code) #HTML status code, 200 is expected as that indicates succescful retrieval

200


In [14]:
#convert the html data into a beautifulsoup object
soup = BeautifulSoup(response.text, 'html.parser')
#specify table object from html response
toronto_table=soup.find('table', {'class':"wikitable"})
toronto_table

<table class="wikitable sortable">
<tbody><tr>
<th>Postal Code
</th>
<th>Borough
</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park, Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor, Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park, Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue, Humber Valley Village
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern, Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3B
</td>
<td

From the output of the previous cell, we can see that the table is being located with the desired Toronto neighborhood data.

In [15]:
#convert BeautifulSoup object to Pandas DataFrame 
df_toronto = pd.read_html(str(toronto_table)) #using Pandas built in function for reading html
df_toronto = pd.DataFrame(df_toronto[0]) #coerce list into DataFrame
print(df_toronto.head()) #verify

  Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront


The output of the cell above demonstrates that the table contains the features that we are looking for (Postal Code, Borough, Neighborhood), however it contains several instances on "Not assigned" data. For this, we will need to clean the data.

Since our goal is to segment Toronto neighborhoods, entry points with no neighborhoods are not useful to us; however, we can substitue the borough in cases where a neighborhood is not assigned. If no borough is available either, it will be necessary to drop the entry.

In [16]:
#Begin cleaning data
NoBoroughIndexes = df_toronto[df_toronto['Borough'] == 'Not assigned'].index #establish indexes where no borough is assigned
df_toronto_bor = df_toronto.drop(NoBoroughIndexes) #drop the indexes obtained
df_toronto_bor.head() #verify succesful drop

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In the output above, we can see that we have succesfully removed rows with no borough assigned, however,we should check to see if there is any data with a borough but no neighborhood.

In [17]:
print(df_toronto_bor[df_toronto_bor['Neighbourhood'] == 'Not assigned'])

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []


We can see above that parsing the neighborhood column for 'Not assigned' entries returns nothing. However, we can double check to make sure there are no rows with unassigned neighborhoods under different entries.

In [18]:
df_neighborhoods = pd.DataFrame(df_toronto_bor['Neighbourhood'].value_counts())
df_neighborhoods

Neighbourhood
Downsview                                                       4
Don Mills                                                       2
Runnymede, Swansea                                              1
Parkview Hill, Woodbine Gardens                                 1
Canada Post Gateway Processing Centre                           1
Woburn                                                          1
Kingsview Village, St. Phillips, Martin Grove G...              1
Woodbine Heights                                                1
St. James Town                                                  1
North Park, Maple Leaf Park, Upwood Park                        1
Victoria Village                                                1
Little Portugal, Trinity                                        1
Studio District                                                 1
Northwest, West Humber - Clairville                             1
Northwood Park, York University                                 1
Dorset Park, Wexford Heights, Scarborough Town ...              1
Church and Wellesley                                            1
Old Mill South, King's Mill Park, Sunnylea, Hum...              1
The Kingsway, Montgomery Road, Old Mill North                   1
Kennedy Park, Ionview, East Birchmount Park                     1
Fairview, Henry Farm, Oriole                                    1
East Toronto, Broadview North (Old East York)                   1
West Deane Park, Princess Gardens, Martin Grove...              1
Cliffside, Cliffcrest, Scarborough Village West                 1
Humberlea, Emery                                                1
Agincourt                                                       1
Golden Mile, Clairlea, Oakridge                                 1
Westmount                                                       1
University of Toronto, Harbord                                  1
The Danforth West, Riverdale                                    1
Mimico NW, The Queensway West, South of Bloor, ...              1
First Canadian Place, Underground city                          1
Caledonia-Fairbanks                                             1
Lawrence Park                                                   1
Central Bay Street                                              1
Islington Avenue, Humber Valley Village                         1
Steeles West, L'Amoreaux West                                   1
Milliken, Agincourt North, Steeles East, L'Amor...              1
Bayview Village                                                 1
Christie                                                        1
Garden District, Ryerson                                        1
Humber Summit                                                   1
Humewood-Cedarvale                                              1
Glencairn                                                       1
Malvern, Rouge                                                  1
Cedarbrae                                                       1
India Bazaar, The Beaches West                                  1
Bedford Park, Lawrence Manor East                               1
Commerce Court, Victoria Hotel                                  1
Richmond, Adelaide, King                                        1
North Toronto West, Lawrence Park                               1
Moore Park, Summerhill East                                     1
New Toronto, Mimico South, Humber Bay Shores                    1
South Steeles, Silverstone, Humbergate, Jamesto...              1
Rosedale                                                        1
Alderwood, Long Branch                                          1
Summerhill West, Rathnelly, South Hill, Forest ...              1
Harbourfront East, Union Station, Toronto Islands               1
CN Tower, King and Spadina, Railway Lands, Harb...              1
Bathurst Manor, Wilson Heights, Downsview North                 1
Del Ray, Mount Dennis, Kee

We can see when looking at the counts for how many times a neighborhood shows up, there are no entries without assigned neighborhoods.
We can also see that many postal areas contain multiple nighborhoods, where neighboroods are separated by commas in the above table.

In [19]:
shape = df_toronto_bor.shape
row = shape[0]
col = shape[1]
fmt_str = "There are %i rows and %i columns in our cleaned data set" % (row, col)
print(fmt_str)

There are 103 rows and 3 columns in our cleaned data set


# Part Two:
## In this section (which will be saved as a new and separate notebook, as per the assignment's instructions), we will find the latitude and longitude data of the previously identified boroughs and neighborhoods

In [23]:
!pip install geocoder

     |████████████████████████████████| 98 kB 10.1 MB/s eta 0:00:01


In [24]:
#To do this, we are going to use geocoder
import geocoder

#we will start with an empty lat/long column
df_toronto_bor['Lat/Long'] = None
df_toronto_bor.reset_index(inplace=True)
df_toronto_bor.drop(['level_0', 'index'], axis=1, inplace=True)

KeyError: "['level_0'] not found in axis"

In [25]:
df_toronto_bor.head()

index Postal Code           Borough  \
0      2         M3A        North York   
1      3         M4A        North York   
2      4         M5A  Downtown Toronto   
3      5         M6A        North York   
4      6         M7A  Downtown Toronto   

                                 Neighbourhood Lat/Long  
0                                    Parkwoods     None  
1                             Victoria Village     None  
2                    Regent Park, Harbourfront     None  
3             Lawrence Manor, Lawrence Heights     None  
4  Queen's Park, Ontario Provincial Government     None

In [29]:
#We can then construct a loop to populate each postal code with a latitude and longitude value

#lat_lng_coords = None

#for code, data in (df_toronto_bor['Postal Code'].items()):
#    while(df_toronto_bor.loc[code, 'Lat/Long'] is None):
#        g = geocoder.google('{}, Toronto, Ontario'.format(data))
#        df_toronto_bor.loc[0, 'Lat/Long'] = g.latlng


KeyboardInterrupt: 

In [30]:
print(lat_lng_coords)
print(g)

None
<[REQUEST_DENIED] Google - Geocode [empty]>


Here we can see that the Geocode library is producing no results, as indicated by the instructions for this project. For this, the loop that would be used to generate the lat/long for each postal code has been commented out.

We will proceed using the CSV file of lat/long values instead

In [31]:
#We will have to import a library, io, to decode the CSV at the link provided
import io
csv_url = 'https://cocl.us/Geospatial_data' #Link to CSV
csv_response = requests.get(csv_url).content #Retrieve content at the link
ll_df = pd.read_csv(io.StringIO(csv_response.decode('utf-8'))) # Decode the content string and coerce it to df
ll_df.head() #Preview df

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [36]:
#drop lat/long column made before failed loop 
df_toronto_bor.drop(['Lat/Long'], axis=1, inplace=True)
#merge original table with lat/long table
neighborhood_ll = pd.merge(df_toronto_bor, ll_df, on=['Postal Code', 'Postal Code'])
neighborhood_ll.drop(['index'], axis=1, inplace=True)
neighborhood_ll.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

Here, we can see that the provided CSV has been uploaded and merged into the main dataframe, so now we have the postal code, corresponding borough, neighborhood, and latitude/longitude values for each.

# Part Three:

## In this section (which will also be saved/published as a new/separate notebook) we will visualize, segment, and cluster the neighborhood data

In [37]:
#import library for clustering
from sklearn.cluster import KMeans
#import library for visualizing geospatial data
import folium

In [48]:
#instantiate a map of the Toronto area with folium
map_toronto = folium.Map(location = [43.66, -79.38], zoom_start = 11) #Using the lat/long values of DT Toronto
map_toronto

In [49]:
#add markers for each neighbnorhood and borough, and redraw the map
for lat, lng, borough, neigh in zip(neighborhood_ll['Latitude'], neighborhood_ll['Longitude'],
                                   neighborhood_ll['Borough'], neighborhood_ll['Neighbourhood']):
    label = '{}, {}'.format(neigh,borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
    [lat,lng],
    radius=5,
    popup=label,
    color='blue',
    fill=True,
    fill_color='#3186cc',
    fill_opacity=0.7,
    parse_html=False).add_to(map_toronto)
    
map_toronto

### Now that we have a labeled map of the Toronto area we can implement Foursquare's API to segment and cluster the area

In [50]:
#Components of the FourSquare API URL
CLIENT_ID='Y2CCZIUQQWKTENZRWWFBDQY3DNOSP3AYUTNDRNY3HYQZ3EV4' #Foursquare ID
CLIENT_SECRET='ZRRPIVBOFR2WTM4XGL0A4OTTGRFGMC0RZJ5QITOMG33UYWFL' #Foursquare Access Secret
VERSION='20180605' #Foursquare API Version
LIMIT=100 #Default Limit

Once we have our credentials ready to pass into the API, we can define a function that will pull those credentials into a URL to access the API. We can then call that function on the neighborhoods in our dataframe, and return the venues in those areas


In [52]:
def getNearbyVenues(names, latitudes, longitudes, radius=500): #Begin function definition, specifying inputs
    
    venues_list=[] #Instantiate empty list to pass venues into
    #Open for loop to pass the name and location of each neighborhood into Foursquare's API
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name) #Progress Tracking that will print each neighborhood as it is processed
        
        #Create the API request URL skeleton
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        lng,
        radius,
        LIMIT)

        #With the URL skeleton established, we can make a GET request that accepts it
        results = requests.get(url).json()["response"]['groups'][0]['items'] #from out lessons, we know this is the location of the data that we are looking for
        
        #We only want some of the information that is produced, and we can define that here
        venues_list.append([(
        name,
        lat,
        lng,
        v['venue']['name'],
        v['venue']['location']['lat'],
        v['venue']['location']['lng'],
        v['venue']['categories'][0]['name']) for v in results])
    
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood',
                    'Neighborhood Latitude',
                    'Neighborhood Longitude',
                    'Venue',
                    'Venue Latitude', 
                    'Venue Longitude',
                    'Venue Category']
    
    return(nearby_venues)

In [53]:
#We can now psas our neighborhood data into this function and generate a new df for the resulting venues
toronto_venues = getNearbyVenues(names=neighborhood_ll['Neighbourhood'],
                                latitudes=neighborhood_ll['Latitude'],
                                longitudes=neighborhood_ll['Longitude']
                                )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Lets check the size of the new dataframe, and preview the first five rows

In [54]:
print(toronto_venues.shape)
toronto_venues.head()

(2141, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3             Tim Hortons       43.725517       -79.313103   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3            Coffee Shop  
4  Portuguese Restaurant

We can also look at the number of venues returned for each neighborhood

In [55]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               5   
Alderwood, Long Branch                                                  7   
Bathurst Manor, Wilson Heights, Downsview North                        21   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      22   
Berczy Park                                                            55   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  13   
Cedarbrae                                                               8   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   75   
Clarks Corners, Tam O'Shanter, Sullivan                                12   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        9   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        5   
Don Mills                                                              23   
Dorset Park, Wexford Heights, Scarborough Town ...                      5   
Downsview                                                              16   
Dufferin, Dovercourt Village                                           13   
East Toronto, Broadview North (Old East York)                           3   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      8   
Fairview, Henry Farm, Oriole                                           70   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Glencairn                                                               5   
Golden Mile, Clairlea, Oakridge                                        10   
Guildwood, Morningside, West Hill                                       8   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
Hillcrest Village                                                       5   
Humber Summit                                                           2   
Humberlea, Emery                                                        1   
Humewood-Cedarvale                                                      5   
India Bazaar, The Beaches West                                         19   
Kennedy Park, Ionview, East Birchmount Park                             4   
Kensington Market, Chinatown, Grange Park                              74   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                                       13   
Lawrence Park                                                           3   
Leaside                                                                33   
Little Portugal, Trinity                                               45   
Malvern, Rouge                                                          1   
Milliken, Agincourt North, Steeles East, L'Amor.

We can transform out data now to a more informative format: we can look at the types of venues present in each neighborhood. To do this, we can encode the venue category, group each neighborhood together, and analyze the average frequency of venue type within each neighborhood.

In [59]:
#We can transform the venue categories into columns of the table. Each row is an entry, and each value is a
#binary result signifying that a category of venue is present at that entry. 
#At this stage, each neighborhood will have one entry per venue, so this is a temporary stage
tor_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
print(tor_onehot.shape)
tor_onehot.head()

(2141, 273)


Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                  0                  0        0                   0   
1                  0                  0        0                   0   
2                  0                  0        0                   0   
3                  0                  0        0                   0   
4                  0                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Stadium  Beach  \
0               0                 0                   0      0   
1               0                 0                   0      0   
2               0                 0                   0      0   
3               0                 0                   0      0   
4               0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0               0        0            0                0         0   
1               0        0            0                0         0   
2               0        0            0                0         0   
3               0        0            0                0         0   
4               0        0            0                0         0   

   Burger Joint  Burrito Place  Bus Line  Bus Station  Business Service  \
0             0    

We can see that in the dataframe above, each row is a particular venue in a particular neighborhood, however, we have lost some information, such as which neighborhood the row represents, so, we can add the neighborhood column back and place it at the beginning of the dataframe

In [63]:
#add neighborhood column
tor_onehot['Neighborhood'] = toronto_venues['Neighborhood']
#move neighborhood column to first position
col_name = 'Neighborhood'
first_col = tor_onehot.pop(col_name)
tor_onehot.insert(0,col_name, first_col)
tor_onehot.head()

Neighborhood  Women's Store  Yoga Studio  Accessories Store  \
0         Parkwoods              0            0                  0   
1         Parkwoods              0            0                  0   
2  Victoria Village              0            0                  0   
3  Victoria Village              0            0                  0   
4  Victoria Village              0            0                  0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0                 0                   0            0              0   
1                 0                   0            0              0   
2                 0                   0            0              0   
3                 0                   0            0              0   
4                 0                   0            0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  \
0           0                   0          0       0              0   
1           0                   0          0       0              0   
2           0                   0          0       0              0   
3           0                   0          0       0              0   
4           0                   0          0       0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bridal Shop  Bubble Tea Shop  Building  Burger Joint  Burri

It is not particularly informative to have a row for every venue in a neighborhood, so we will group the table by neighborhood and take the average frequency of each venue category. This will give a more general picture of what types of venues appear in which neighborhoods

In [64]:
tor_grp = tor_onehot.groupby('Neighborhood').mean().reset_index()
tor_grp

Neighborhood  Women's Store  \
0                                           Agincourt       0.000000   
1                              Alderwood, Long Branch       0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North       0.000000   
3                                     Bayview Village       0.000000   
4                   Bedford Park, Lawrence Manor East       0.000000   
5                                         Berczy Park       0.000000   
6                         Birch Cliff, Cliffside West       0.000000   
7        Brockton, Parkdale Village, Exhibition Place       0.000000   
8   Business reply mail Processing Centre, South C...       0.000000   
9   CN Tower, King and Spadina, Railway Lands, Har...       0.000000   
10                                Caledonia-Fairbanks       0.250000   
11              Canada Post Gateway Processing Centre       0.000000   
12                                          Cedarbrae       0.000000   
13                                 Central Bay Street       0.000000   
14                                           Christie       0.000000   
15                               Church and Wellesley       0.000000   
16            Clarks Corners, Tam O'Shanter, Sullivan       0.000000   
17    Cliffside, Cliffcrest, Scarborough Village West       0.000000   
18                     Commerce Court, Victoria Hotel       0.000000   
19                                         Davisville       0.000000   
20                                   Davisville North       0.000000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn       0.000000   
22                                          Don Mills       0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...       0.000000   
24                                          Downsview       0.000000   
25                       Dufferin, Dovercourt Village       0.000000   
26      East Toronto, Broadview North (Old East York)       0.000000   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...       0.000000   
28                       Fairview, Henry Farm, Oriole       0.042857   
29             First Canadian Place, Underground city       0.000000   
30    Forest Hill North & West, Forest Hill Road Park       0.000000   
31                           Garden District, Ryerson       0.000000   
32                                          Glencairn       0.000000   
33                    Golden Mile, Clairlea, Oakridge       0.000000   
34                  Guildwood, Morningside, West Hill       0.000000   
35  Harbourfront East, Union Station, Toronto Islands       0.000000   
36                      High Park, The Junction South       0.000000   
37                                  Hillcrest Village       0.000000   
38                                      Humber Summit       0.000000   
39                                   Humberlea, Emery       0.000000   
40                                 Humewood-Cedarvale       0.000000   
41                     India Bazaar, The Beaches West       0.000000   
42        Kennedy Park, Ionview, East Birchmount Park       0.000000   
43          Kensington Market, Chinatown, Grange Park       0.000000   
44  Kingsview Village, St. Phillips, Martin Grove ...       0.000000   
45                   Lawrence Manor, Lawrence Heights       0.076923   
46                                      Lawrence Park       0.000000   
47                                            Leaside       0.000000   
48                           Little Portugal, Trinity       0.000000   
49                                     Malvern, Rouge       0.000000   
50  Milliken, Agincourt North, Steeles East, L'Amo...       0.000000   
51  Mimico NW, The Queensway West, South of Bloor,...       0.000000   
52                        Moore Park, Summerhill East       0.000000   
53       New Toronto, Mimico South, Humber Bay Shores       0.000000   
54           North Park, Maple Leaf Park, Upwood Park       0.000000   
55       

In [65]:
tor_grp.shape

(96, 273)

This demonstrates that we have 96 neighborhoods with 273 venue types. We can also see that it is often the case that a venue does not show up in a neighborhood at all, or very seldom, so it may be more informative to generate the top five frequent venues for each neighborhood.

In [66]:
num_top_venues = 5

for hood in tor_grp['Neighborhood']:
    print("---"+hood+"---")
    temp = tor_grp[tor_grp['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp=temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp=temp.round({'freq':2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

---Agincourt---
                       venue  freq
0             Clothing Store   0.2
1                     Lounge   0.2
2               Skating Rink   0.2
3             Breakfast Spot   0.2
4  Latin American Restaurant   0.2


---Alderwood, Long Branch---
            venue  freq
0     Pizza Place  0.29
1             Gym  0.14
2     Coffee Shop  0.14
3             Pub  0.14
4  Sandwich Place  0.14


---Bathurst Manor, Wilson Heights, Downsview North---
           venue  freq
0           Bank  0.10
1    Coffee Shop  0.10
2    Gas Station  0.05
3          Diner  0.05
4  Shopping Mall  0.05


---Bayview Village---
                 venue  freq
0                 Café  0.25
1  Japanese Restaurant  0.25
2   Chinese Restaurant  0.25
3                 Bank  0.25
4  Monument / Landmark  0.00


---Bedford Park, Lawrence Manor East---
                venue  freq
0  Italian Restaurant  0.09
1         Coffee Shop  0.09
2      Sandwich Place  0.09
3   Indian Restaurant  0.05
4           Locksmith  0.

Here we can see a visual overview of the top five most frequent venue types within a neighborhood, however, we cannot manipulate this data until we store it in a dataframe. So, we should sort them in descending order, and then place the data into a new df.

In [70]:
#import library to return evenly spaced values within interval
import numpy as np
#it will be necessary to define a new function specific to the data we want to sort
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

indicators = ['st', 'nd', 'rd'] #unique indicators for 1st, 2nd, 3rd
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind])) #if 1st, 2nd, 3rd
    except:
        columns.append('{}th Most Common Venue'.format(ind+1)) #anything greater than 3rd
        
#instantiate new df
n_venues_sorted = pd.DataFrame(columns=columns)
n_venues_sorted['Neighborhood'] = tor_grp['Neighborhood']

#loop for each neighborhood in grouped neighborhood df to be passed through most common venue function
for ind in np.arange(tor_grp.shape[0]):
    n_venues_sorted.iloc[ind,1:] = return_most_common_venues(tor_grp.iloc[ind, :], num_top_venues)
    
n_venues_sorted.head()                                                       

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0          Skating Rink  Latin American Restaurant        Clothing Store   
1                   Gym                Coffee Shop              Pharmacy   
2                  Bank             Ice Cream Shop    Frozen Yogurt Shop   
3   Japanese Restaurant         Chinese Restaurant                  Bank   
4    Italian Restaurant             Sandwich Place   American Restaurant   

  5th Most Common Venue  
0        Breakfast Spot  
1                   Pub  
2        Sandwich Place  
3          Dessert Shop  
4                   Pub

# Clustering
### Now that the neighborhoods can be characterized by their most common venues, we can perform k-means clustering on the neighborhoods

In [71]:
#we will set the number of clusters arbitrarily to 5
kclusters = 5

tor_grp_cls = tor_grp.drop('Neighborhood',1)
kmeans = KMeans(n_clusters = kclusters, random_state=0).fit(tor_grp_cls)

#check labels
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [102]:
#We can import the cluster identification into our dataframe with most common venues
n_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
#Create new df to aggreate all data
tor_merge = neighborhood_ll
#rename to match spellings across df
tor_merge.rename({'Neighbourhood':'Neighborhood'}, axis=1, inplace=True)

#merge df's on the neighborhoods
tor_merge = tor_merge.join(n_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
tor_merge.head(10)

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village   
6         M1B       Scarborough                               Malvern, Rouge   
7         M3B        North York                                    Don Mills   
8         M4B         East York              Parkview Hill, Woodbine Gardens   
9         M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0     Food & Drink Shop   
1  43.725882 -79.315572             1.0          Intersection   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   
5  43.667856 -79.532242             NaN                   NaN   
6  43.806686 -79.194353             4.0  Fast Food Restaurant   
7  43.745906 -79.352188             1.0                   Gym   
8  43.706397 -79.309937             1.0           Pizza Place   
9  43.657162 -79.378937             1.0        Clothing Store   

  2nd Most Common Venue   3rd Most Common Venue  4th Most Common Venue  \
0                  Park             Wings Joint     Dim Sum Restaurant   
1     French Restaurant             Pizza Place            Coffee Shop   
2                  Park                     Pub                 Bakery   
3         Women's Store  Furniture / Home Store  Vietnamese Restaurant   
4           Art Gallery      Italian Restaurant               Beer Bar   
5                   NaN                     NaN                    NaN   
6            Donut Shop      Dim Sum Restaurant                  Diner   
7   Japanese Restaurant             Coffee Shop             Beer Store   
8        Breakfast Spot    Gym / Fitness Center              Pet Store   
9           Coffee Shop                    Café        Bubble Tea Shop   

  5th Most Common Venue  
0                 Diner  
1          Hockey Arena  
2               Theater  
3     Accessories Store  
4                   Bar  
5                   NaN  
6        Discount Store  
7    Dim Sum Restaurant  
8              Pharmacy  
9   Japanese Restaurant

We can see from the preview of the new dataframe that some neighborhoods were not clustered because they had no
venues present. For the purposes of this project, we can drop those rows from the table.

In [103]:
#drop rows with NA values
tor_merge.dropna(axis=0,inplace=True)

### With the top five venues per neighborhood identifed, each neighborhood assigned to a cluster, and the lat/long of each neighborhood provided, we can visualize the result of our clustering

In [105]:
tor_merge.head(10)
print(tor_merge.dtypes)

Postal Code               object
Borough                   object
Neighborhood              object
Latitude                 float64
Longitude                float64
Cluster Labels           float64
1st Most Common Venue     object
2nd Most Common Venue     object
3rd Most Common Venue     object
4th Most Common Venue     object
5th Most Common Venue     object
dtype: object


In [106]:
#Coerce Cluster Lables to int
tor_merge['Cluster Labels'] = tor_merge['Cluster Labels'].astype(int)
print(tor_merge.dtypes)

Postal Code               object
Borough                   object
Neighborhood              object
Latitude                 float64
Longitude                float64
Cluster Labels             int64
1st Most Common Venue     object
2nd Most Common Venue     object
3rd Most Common Venue     object
4th Most Common Venue     object
5th Most Common Venue     object
dtype: object


In [108]:
#create map
map_cls = folium.Map(location=[43.66, -79.38], zoom_start = 11)

#set cluster colors
import matplotlib.cm as cm
import matplotlib.colors as colors
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0,1,len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

#add markers
markers_colors = []
for lat, lon, poi, cluster in zip(tor_merge['Latitude'], tor_merge['Longitude'], tor_merge['Neighborhood'], tor_merge['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
    [lat,lon],
    radius=5,
    popup=label,
    color=rainbow[cluster-1],
    fill=True,
    fill_color=rainbow[cluster-1],
    fill_opacity=0.7).add_to(map_cls)
    
map_cls

We can now compare this map to the characteristics of each cluster

# Cluster 1

In [122]:
tor_merge.drop(['Borough', 'Postal Code'], axis=1, inplace = True)
tor_merge.loc[tor_merge['Cluster Labels'] == 0, tor_merge.columns[[0] +list(range(4,tor_merge.shape[1]))]]

Neighborhood 1st Most Common Venue  \
0                                           Parkwoods     Food & Drink Shop   
21                                Caledonia-Fairbanks                  Park   
35      East Toronto, Broadview North (Old East York)                  Park   
49           North Park, Maple Leaf Park, Upwood Park                  Park   
61                                      Lawrence Park                  Park   
64                                             Weston                  Park   
66                                    York Mills West     Convenience Store   
85  Milliken, Agincourt North, Steeles East, L'Amo...          Intersection   
91                                           Rosedale                  Park   

   2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
0                   Park                 Wings Joint    Dim Sum Restaurant   
21                  Pool               Women's Store         Grocery Store   
35          Intersection           Convenience Store           Escape Room   
49                Bakery  Construction & Landscaping            Donut Shop   
61              Bus Line                 Swim School           Escape Room   
64           Wings Joint                  Donut Shop    Dim Sum Restaurant   
66                  Park                  Donut Shop    Dim Sum Restaurant   
85                  Park                  Playground                Bakery   
91            Playground                       Trail      Doner Restaurant   

   5th Most Common Venue  
0                  Diner  
21          Dance Studio  
35     Electronics Store  
49                 Diner  
61     Electronics Store  
64                 Diner  
66                 Diner  
85      Doner Restaurant  
91          Dessert Shop

# Cluster 2

In [123]:
tor_merge.loc[tor_merge['Cluster Labels'] == 1, tor_merge.columns[[0] +list(range(4,tor_merge.shape[1]))]]

Neighborhood 1st Most Common Venue  \
1                                     Victoria Village          Intersection   
2                            Regent Park, Harbourfront           Coffee Shop   
3                     Lawrence Manor, Lawrence Heights        Clothing Store   
4          Queen's Park, Ontario Provincial Government           Coffee Shop   
7                                            Don Mills                   Gym   
8                      Parkview Hill, Woodbine Gardens           Pizza Place   
9                             Garden District, Ryerson        Clothing Store   
10                                           Glencairn                  Park   
12              Rouge Hill, Port Union, Highland Creek                   Bar   
13                                           Don Mills                   Gym   
14                                    Woodbine Heights          Skating Rink   
15                                      St. James Town           Coffee Shop   
16                                  Humewood-Cedarvale          Hockey Arena   
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...     Convenience Store   
18                   Guildwood, Morningside, West Hill   Rental Car Location   
19                                         The Beaches     Health Food Store   
20                                         Berczy Park           Coffee Shop   
22                                              Woburn           Coffee Shop   
23                                             Leaside   Sporting Goods Shop   
24                                  Central Bay Street           Coffee Shop   
25                                            Christie         Grocery Store   
26                                           Cedarbrae    Athletics & Sports   
27                                   Hillcrest Village                  Pool   
28     Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
29                                    Thorncliffe Park        Sandwich Place   
30                            Richmond, Adelaide, King           Coffee Shop   
31                        Dufferin, Dovercourt Village                Bakery   
32                                 Scarborough Village            Smoke Shop   
33                        Fairview, Henry Farm, Oriole        Clothing Store   
34                     Northwood Park, York University    Miscellaneous Shop   
36   Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
37                            Little Portugal, Trinity                   Bar   
38         Kennedy Park, Ionview, East Birchmount Park           Coffee Shop   
39                                     Bayview Village                  Café   
40                                           Downsview         Grocery Store   
41                        The Danforth West, Riverdale      Greek Restaurant   
42            Toronto Dominion Centre, Design Exchange           Coffee Shop   
43        Brockton, Parkdale Village, Exhibition Place                  Café   
44                     Golden Mile, Clairlea, Oakridge                Bakery   
46                                           Downsview         Grocery Store   
47                      India Bazaar, The Beaches West                  Park   
48                      Commerce Court, Victoria Hotel           Coffee Shop   
50                                       Humber Summit           Pizza Place   
51     Cliffside, Cliffcrest, Scarborough Village West                 Motel   
53                                           Downsview         Grocery Store   
54                                     Studio District           Coffee Shop   
55                   Bedford Park, Lawrence Manor East           Coffee Shop   
56    Del Ray, Mount Dennis, Keelsdale and Silverthorn                   Bar   
57                                    Humberlea, Emery        Baseball Field   
58                         Birch Cliff, Cl

# Cluster 3

In [124]:
tor_merge.loc[tor_merge['Cluster Labels'] == 2, tor_merge.columns[[0] +list(range(4,tor_merge.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
45  York Mills, Silver Hills   Martial Arts School           Wings Joint   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
45            Donut Shop    Dim Sum Restaurant                 Diner

# Cluster 4

In [125]:
tor_merge.loc[tor_merge['Cluster Labels'] == 3, tor_merge.columns[[0] +list(range(4,tor_merge.shape[1]))]]

Neighborhood 1st Most Common Venue  \
11  West Deane Park, Princess Gardens, Martin Grov...            Print Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
11           Wings Joint      Doner Restaurant    Dim Sum Restaurant   

   5th Most Common Venue  
11                 Diner

# Cluster 5

In [126]:
tor_merge.loc[tor_merge['Cluster Labels'] == 4, tor_merge.columns[[0] +list(range(4,tor_merge.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
6  Malvern, Rouge  Fast Food Restaurant            Donut Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
6    Dim Sum Restaurant                 Diner        Discount Store

# We can see, with two very large clusters and three very small clusters, that our K may not have been an optimal selection. It is likely that we could derive more clusters from cluster 1 and 2, however, the project demonstrates effective clustering and geospatial visualize regardless.